src/model_theory/semantics.lean

```diff
@@ -304,7 +312,29 @@ begin
   rw [if_pos i.is_lt],
 end
 
-lemma realize_all_lift_at_one_self [nonempty M] {n : ℕ} {φ : L.bounded_formula α n}
+lemma realize_subst {φ : L.bounded_formula α n} {tf : α → L.term β} {v : β → M} {xs : fin n → M} :
+  (φ.subst tf).realize v xs ↔ φ.realize (λ a, (tf a).realize v) xs :=
```

Comment:
I believe the three `simp`s in this proof are all the same equality. Maybe prove that first as a separate lemma or `have` statement, that should probably also avoid the use of the `congr`.

Note that the tactic `rcongr` exists to recursively apply `congr'` and `(fun)ext` for you. Both occurrences of `rw iff_eq_eq / refine congr` can be replaced by this tactic (hopefully that is not necessary anymore after using the above suggestion, but maybe nice for future use)